src/Strategy/Python/Poetry/Common.hs

```diff
@@ -0,0 +1,160 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Strategy.Python.Poetry.Common (
+  PoetryProject (..),
+  buildPyProjectGraph,
+  buildGraphWithLock,
+  buildPackageNameGraph,
+  discover,
+) where
+
+import Control.Effect.Diagnostics (Diagnostics, context)
+import Data.Map.Strict qualified as M
+import Data.Maybe (fromMaybe)
+import Data.Text (toLower)
+import DepTypes
+import Discovery.Walk (
+  WalkStep (WalkContinue, WalkSkipAll),
+  findFileNamed,
+  walk',
+ )
+import Effect.Exec
+import Effect.ReadFS (ReadFS, readContentsToml)
+import Graphing (Graphing, addNode, edge, empty, fromList, gmap, promoteToDirect)
+import Path (Abs, Dir, File, Path)
+import Strategy.Python.Poetry.PoetryLock (PackageName (..), PoetryLock (..), PoetryLockPackage (..), poetryLockCodec, toMap)
+import Strategy.Python.Poetry.PyProject (PyProject (..), getDependencies, pyProjectCodec, usesPoetryBuildSystem)
+import Types (DiscoveredProject (..))
+
+newtype PyProjectTomlFile = PyProjectTomlFile {pyProjectTomlPath :: Path Abs File} deriving (Eq, Ord, Show)
+newtype PoetryLockFile = PoetryLockFile {poetryLockPath :: Path Abs File} deriving (Eq, Ord, Show)
+newtype ProjectDir = ProjectDir {pyProjectPath :: Path Abs Dir} deriving (Eq, Ord, Show)
+
+data PoetryProject
+  = PoetryPyProjectOnly ProjectDir PyProjectTomlFile
+  | PoetryPyProjectWithLock ProjectDir PyProjectTomlFile PoetryLockFile
+  deriving (Eq, Ord, Show)
+
+discover :: (Has ReadFS sig m, Has Diagnostics sig m, Has ReadFS rsig run, Has Diagnostics rsig run) => Path Abs Dir -> m [DiscoveredProject run]
+discover dir = context "Poetry" $ do
+  projects <- context "Finding projects" $ findProjects dir
+  pure (map mkProject projects)
+
+findProjects :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> m [PoetryProject]
+findProjects = walk' $ \dir _ files -> do
+  let poetryLockFile = findFileNamed "poetry.lock" files
+  let pyprojectFile = findFileNamed "pyproject.toml" files
+
+  case (poetryLockFile, pyprojectFile) of
+    (Nothing, Just pyproject) -> do
+      poetryProject <- readContentsToml pyProjectCodec pyproject
+      if usesPoetryBuildSystem poetryProject
+        then pure ([PoetryPyProjectOnly (ProjectDir dir) (PyProjectTomlFile pyproject)], WalkSkipAll)
+        else pure ([], WalkContinue)
+    (Just poetrylock, Just pyproject) ->
+      pure
+        (
+          [ PoetryPyProjectWithLock (ProjectDir dir) (PyProjectTomlFile pyproject) (PoetryLockFile poetrylock)
+          ]
+        , WalkSkipAll
+        )
+    -- Without pyproject file, it is unlikely that project is a poetry package. Poetry itself does not work
+    -- without pyproject.toml manifest.
```

Comment:
For context:

in general, lockfiles are produced by project description files.

* `Cargo.toml` -> `Cargo.lock`
* `package.json` -> `yarn.lock` (if yarn is used)
* `Pipfile` -> `Pipfile.lock`

I would say that in general, unless you find a project description file, you aren't actually looking at a real project.  If there are counterexamples, I would call those exceptions, rather than the rule.  In poetry, you cannot run project installation without the `pyproject.toml` file, even if you have a `poetry.lock` file, so this fits the rule.  

There are not going to be many projects that have poetry lock files without a `pyproject.toml`, but if they do, we can probably be sure that they are not using poetry as their build tool.  The one likely exception is that they are running our tool away from their normal builds, like in a specialized container for running fossa, that doesn't copy the `pyproject.toml` file into the container.  In that case, we should simply have them include the `pyproject.toml` file anyway, since it will give us better results than if we tried to just parse the `poetry.lock` file alone (which is technically possible).

@meghfossa, I suggest finding docs that support the above information as the best way to illustrate this decision.

Reply:
Yeah, I understand the general relationship between manifest files and lockfiles. I'm just surprised you thought of this case @meghfossa - I assumed that you came across an actual example of this. If so, we should document when this could happen. I just want to make sure I'm not missing something special and weird about `pyproject.toml`.

Reply:
I haven't come across any example of this in wild, but I wanted to be conscious of all combinations that can possibly happen - so if by any chance, we do come across this scenario, there is debug msg and explicit branch to delineate. I will add most relevant poetry docs link for context. I can add test case in integration tests, but other than that not sure what else I can add.

Reply:
No worries - I just wanted to make sure I wasn't missing some weird Poetry-specific convention.